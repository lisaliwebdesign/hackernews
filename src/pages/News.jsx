import React, { Fragment } from "react";
import { inject, observer } from "mobx-react";
import InfiniteScroll from "react-infinite-scroller";
import { PageContainer } from "../elements/common";
import Nav from "../components/Nav";
import NewsItem from "../components/NewsItem";

@inject("NewsStore")
@observer
export default class News extends React.Component {
  //Page onload
  constructor(props) {
    super(props);
    this.state = {
      hasMoreItems: true,
      threshold: 250,
      pathId: ""
    };
  }
  componentDidMount() {
    //Get path from router
    const path = this.props.location.pathname.split("/")[2];
    //Set load path and load item = 30
    this.props.NewsStore.loadData(path, 30);
    this.setState({
      pathId: path
    });
  }
  componentWillReceiveProps(nextProps) {
    //Update path from props
    const path = nextProps.location.pathname.split("/")[2];
    //Update load path and load item = 30
    this.props.NewsStore.loadData(path, 30);
    this.setState({
      pathId: path
    });
    this.loadItems.bind(this);
  }

  //InfiniteScroll loading
  loadItems(page) {
    var self = this;
    //Get stored data
    const NewsStore = this.props.NewsStore;
    const newsArrarySet = NewsStore.newsDataGroupArray;
    const newsArrarySetCount = NewsStore.newsDataGroupArrayCount;
    //Load more page when the request is less than the tatal nunber of data set
    if (page < newsArrarySetCount + 1) {
      NewsStore.loadItems(newsArrarySet[page]);
    } else {
      self.setState({
        hasMoreItems: false
      });
    }
  }

  renderContent = text => (
    <Fragment>
      <Nav selectedId={this.state.pathId} /> <PageContainer content={text} />
    </Fragment>
  );

  newsContent = data => {
    const content = <NewsItem data={data} type={this.state.pathId} />;
    return this.renderContent(content);
  };

  render() {
    const { renderContent, newsContent } = this;
    const NewsStore = this.props.NewsStore;
    if (NewsStore.loadDataError) {
      return renderContent("Sorry... very embarassing");
    }
    if (!NewsStore.newsItemsData || NewsStore.newsItemsData.length === 0) {
      return renderContent("Loading...");
    }
    if (NewsStore.newsItemsData) {
      return (
        <InfiniteScroll
          pageStart={0}
          loadMore={this.loadItems.bind(this)}
          hasMore={this.state.hasMoreItems}
          threshold={this.state.threshold}
          loader={
            <div className="loader" key={0}>
              Loading ...
            </div>
          }
        >
          {newsContent(NewsStore.newsItemsData)}
        </InfiniteScroll>
      );
    }
  }
}
