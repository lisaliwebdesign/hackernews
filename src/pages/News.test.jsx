import News from "./News";
import { Link } from "react-router";
import { Provider } from "mobx-react";
import { MemoryRouter as Router, withRouter } from "react-router-dom"; // 4.0.0

const testHelp = {
  mountComponent: (props = {}) => {
    const moutWrapper = mount(
      <Provider NewsStore={props.NewsStore}>
        <Router>
          <News router={props.Nav} />
        </Router>
      </Provider>
    );
    return moutWrapper;
  }
};

it("<Pages><News - renders component", () => {
  const NewsStore = { loadDataError: true };
  const Nav = mount(<a>TEST</a>);
  const wrapper = render(
    <Provider NewsStore={NewsStore}>
      <Router>
        <News router={Nav} />
      </Router>
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
});

it("<components><News> - with error data", () => {
  const NewsStore = { loadDataError: true };
  //mount
  const Nav = mount(<a>TEST</a>);
  //const Path = mount("");
  const wrapper = render(
    <Provider NewsStore={NewsStore}>
      <Router>
        <News router={Nav} location={{ pathname: "Path" }} />
      </Router>
    </Provider>
  );

  expect(wrapper.find("div").hasClass("container")).toBe(true);
});
