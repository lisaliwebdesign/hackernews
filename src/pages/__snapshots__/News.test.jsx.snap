// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pages><News - renders component 1`] = `
Array [
  <nav
    aria-label="main navigation"
    class="navbar navbar-guest"
>
    <div
        class="container"
    >
        <div
            class="navbar-brand"
        >
            <a
                class="navbar-item"
                href="/"
            >
                <h1
                    class="title has-text-centered is-2 hp-title"
                >
                    Hacker News
                </h1>
            </a>
            <button
                class="button navbar-burger"
                data-target="navMenu"
            >
                <span>
                    
                </span>
                 
                <span>
                    
                </span>
                 
                <span>
                    
                </span>
            </button>
        </div>
        <div
            class="navbar-menu"
            id="navMenu"
        >
            <div
                class="navbar-start"
            >
                <a
                    class="navbar-item active"
                    href="/news/topstories"
                    id="topstories"
                >
                    Top stories
                </a>
                <a
                    class="navbar-item"
                    href="/news/newstories"
                    id="newstories"
                >
                    New stories
                </a>
                <a
                    class="navbar-item"
                    href="/news/beststories"
                    id="beststories"
                >
                    Comments
                </a>
                <a
                    class="navbar-item"
                    href="/news/askstories"
                    id="askstories"
                >
                    Ask
                </a>
                <a
                    class="navbar-item"
                    href="/news/showstories"
                    id="showstories"
                >
                    Show
                </a>
                <a
                    class="navbar-item"
                    href="/news/jobstories"
                    id="jobstories"
                >
                    Job
                </a>
            </div>
            <div
                class="navbar-end"
            >
                <div
                    class="navbar-item"
                >
                    <a
                        href="/user/login"
                    >
                        SIGN IN
                    </a>
                </div>
            </div>
        </div>
    </div>
</nav>,
  " ",
  <div
    class="container"
>
    Sorry... very embarassing
</div>,
]
`;
